Prototype/fdb layer replace couch rate

davisp opened the pull request:
## Overview

Replace the couch_rate rate limiting algorithm with couch_views_batch to dynamically optimize the indexer batch size during view updates.

While working on optimizing couch_views I ran into couch_rate behaving oddly. Reading through the implementation it was nearly impossible to predict how changing various parameters would affect the behavior under load. The rate limiting algorithms were also a bit misapplied. We don't really want a rate "limiter" so much as a rate "maximizer". 

After failing to comprehend couch_rate and realizing that it was missing some fairly important signals (specifically, the approximate transaction size) I decided to take a whack at simplifying things. The new approach is quite a bit simpler and results in the ability to both successfully index large views (1,000,000 doc ranges were tested) but also has a more predictable and simple behavior.

![batch_size_search](https://user-images.githubusercontent.com/19929/92163250-f6697480-edf8-11ea-9553-9da2be9f1152.png)

The graph above is a representative of the behavior. Batch sizes start out small when a view indexer starts. The batch size is quickly ramped up to search for the threshold of a batch size, and then batch sizes are adjusted in much smaller increments to follow the optimal batch size.

Matching the old behavior of couch_rate to attempt to maintain a consistent transaction time limit can be accomplished trivially by adjusting the `couch_views.batch_tx_max_time` parameter.

## Testing recommendations

`make check`

## Checklist

- [x] Code is written and works correctly
- [x] Changes are covered by tests
- [x] Any new configurable parameters are documented in `rel/overlay/etc/default.ini`
- [ ] A PR for documentation changes has been made in https://github.com/apache/couchdb-documentation

davisp commented:
I should mention that my hack to generate the batch_size graphs had a bug where it would double count if a transaction would be retried. that spike to ~8500 just happens to be at the transition in this graph. There's always a conflict towards the start of view indexing due to a conflict on the `couch_jobs` data that I have yet to track down.

garrensmith commented on src/couch_views/src/couch_views_batch.erl:
```diff
@@ -0,0 +1,192 @@
+    }.
+
+
+-spec failure(State::term()) -> NewState::term().
```
Can you also make these specs more specific

garrensmith commented on src/couch_views/src/couch_views_indexer.erl:
```diff
@@ -165,48 +171,60 @@ add_error(Error, Reason, Data) ->
+        end
+    catch
+        error:{erlfdb_error, Error} ->
+            case lists:member(Error, ?RETRY_FAILURES) of
```
This is really nice.

garrensmith reviewed (changes requested):
In general, this looks really good and is a lot simpler than the previous implementation. 
I'm not convinced about having the default implementation of `batch_module` in the same file as `couch_views_batch`. I initially found it confusing trying to figure out where it starts and where `couch_views_batch` is. I would prefer we have a `couch_batch` module similar to `couch_rate` with the default module and `couch_views_batch`.

iilyak commented on src/couch_views/src/couch_views_batch.erl:
```diff
@@ -0,0 +1,192 @@
+]).
+
+
+-callback start(State::term()) -> {NewState::term(), BatchSize::pos_integer()}.
```
Should we define and use `-opaque state() :: term().`?

davisp replied:
I'm not sure what that buys us or how it'd interact with other modules implementing the behavior?

iilyak replied:
With you latest rewrite and separation into two modules it is not required to use `opaque`.

iilyak commented on src/couch_views/src/couch_views_batch.erl:
```diff
@@ -0,0 +1,192 @@
+    get(?MODULE).
+
+
+save_state(Mod, Batch) ->
```
Should we rename `Batch` into `State`?

davisp replied:
Done.

iilyak commented on src/couch_views/src/couch_views_batch.erl:
```diff
@@ -0,0 +1,192 @@
+    search_incr,
+    sense_incr,
+    max_tx_size,
+    max_tx_time,
```
`max_tx_time_msec` would be better.

davisp replied:
Done.

iilyak commented on src/couch_views/src/couch_views_batch.erl:
```diff
@@ -0,0 +1,192 @@
+        size = get_config("batch_initial_size", "100"),
+        search_incr = get_config("batch_search_increment", "500"),
+        sense_incr = get_config("batch_sense_increment", "100"),
+        max_tx_size = get_config("batch_max_tx_size", "9000000"),
```
`max_tx_time_msec` would be better

iilyak commented on src/couch_views/src/couch_views_batch.erl:
```diff
@@ -0,0 +1,192 @@
+        search_incr = get_config("batch_search_increment", "500"),
+        sense_incr = get_config("batch_sense_increment", "100"),
+        max_tx_size = get_config("batch_max_tx_size", "9000000"),
+        max_tx_time = get_config("batch_max_tx_time", "4500"),
```
`max_tx_time_msec` would be better

davisp replied:
Done.

iilyak commented on src/couch_views/README.md:
```diff
@@ -16,3 +17,18 @@ Code layout:
+; batch_search_increment = 500 ; Size change when searching for the threshold
+; batch_sense_increment = 100 ; Size change increment after hitting a threshold
+; batch_max_tx_size = 9000000 ; Maximum transaction size in bytes
+; batch_max_tx_time = 4500 ; Maximum transaction time in milliseconds
```
`max_tx_time_msec` would be better

davisp replied:
Done.

iilyak commented on rel/overlay/etc/default.ini:
```diff
@@ -309,6 +309,14 @@ iterations = 10 ; iterations for password hashing
+; batch_search_increment = 500 ; Size change when searching for the threshold
+; batch_sense_increment = 100 ; Size change increment after hitting a threshold
+; batch_max_tx_size = 9000000 ; Maximum transaction size in bytes
+; batch_max_tx_time = 4500 ; Maximum transaction time in milliseconds
```
`batch_max_tx_time_msec` would be better

davisp replied:
Done.

iilyak commented:
> I'm not convinced about having the default implementation of batch_module in the same file as couch_views_batch. I initially found it confusing trying to figure out where it starts and where couch_views_batch is. I would prefer we have a couch_batch module similar to couch_rate with the default module and couch_views_batch.

I agree with Garren. This separation would also hide the `#batch_st{}` record in `couch_views_batch` module.

iilyak commented:
Just wanted to mention that the scope of `couch_rate` was bigger. The idea behind `couch_rate` was to provide a `rate limiter`/`batch size estimator` which can be used anywhere where we have a chance of a transaction failure from FDB and where we can be better off by adjusting transaction size (one possible use case is to unify with https://github.com/apache/couchdb/blob/prototype/fdb-layer/src/couch_replicator/src/couch_replicator_rate_limiter.erl). However since it seems like we are getting [a consensus that we should introduce limits](https://lists.apache.org/thread.html/rac6c90c4ae03dc055c7e8be6eca1c1e173cf2f98d2afe6d018e62d29%40%3Cdev.couchdb.apache.org%3E) we might not need a generic solution and `couch_view` specific one might suffice.

iilyak commented:
In regards to `Move error reporting test to EUnit` commit you've mentioned: 

> This test doesn't fail correctly any longer. Rather than attempting to
create a new pathological view case I've just moved it to eunit where we can use meck to throw errors directly.

The meck can be used from ExUnit. 
```
  defp with_fdb_error(_ctx) do
    :ok = :meck.new(:couch_views_batch, [:passthrough])
    :meck.expect(:couch_views_batch, :start, [], fn() ->
      :erlang.error({:erlfdb_error, 2101})
    end)
    on_exit(fn ->
      :meck.unload()
    end)
  end

describe "something something" do
    setup [....,  :with_fdb_error]
    test "something something", ctx do
       ...
    end
end
```

davisp commented on src/couch_views/src/couch_views_batch.erl:
```diff
@@ -0,0 +1,192 @@
+        TxSize::non_neg_integer(),
+        DocsRead::non_neg_integer(),
+        State::term()
+    ) -> NewState::term().
```
These specs match the callbacks of the behavior for couch_views_batch so I'm not sure whether that's appropriate or not?

nickva commented on src/couch_views/src/couch_views_indexer.erl:
```diff
@@ -165,48 +171,60 @@ add_error(Error, Reason, Data) ->
+                    couch_views_batch:failure(),
+                    update(Db, Mrst0, State0);
+                false ->
+                    erlang:error({erlfdb_error, Error})
```
Will this loose the original stacktrace? It might not be as nice but we could defined ?RETRY_FAILURES as a guard maybe. Would this work:

```erlang
-define(RETRY_FAILURE(E), (E =:= 1007 orelse E =:= 1031 ...)).
```
And then say `catch error:{erlfdb_error, E} when ?RETRY_FAILURE(E) ...` ?

davisp replied:
That seems reasonable. I'll update to use that pattern.

nickva commented on src/couch_views/src/couch_views_indexer.erl:
```diff
@@ -165,48 +171,60 @@ add_error(Error, Reason, Data) ->
+                    tx_size := TxSize
+                } = State1,
+                couch_views_batch:success(TxSize, DocsRead),
+                update(Db, Mrst1, State1)
```
Because of the catch that wraps it, would this still be tail recursive? Say if we call this 1M times would it have to keep the stack frames around to unwind after the call?

davisp replied:
Ooooh, good catch. Will fix that.

davisp commented:
@garrensmith I've split the behavior/implementation between `couch_views_batch` (behavior) and `couch_views_batch_impl` (implementation). I don't believe this behavior is nearly abstract enough to cover anything besides couch_views so it doesn't make a lot of sense to me to create a new application to hold a new module that only couch_views will use.

@iilyak For the scope I agree that this is a lot more narrow. I started by trying to figure out the AIMD/congestion algorithms in both couch_rate and couch_replicator but after thinking awhile I realized that having an AIMD based algorithm isn't appropriate for batch sizes since we want to avoid failing batches as much as possible because every failed batch is wasted time when building a view. I did a number of different approaches to try and measure optimal batch sizes and behaviors. An earlier version of this work actually tried to optimize the rate of indexing within the bounds of transaction sizes. That lead to the realization that just maximizing the work done in a single transaction was most beneficial overall for indexing throughput. That's how it ended up being so simple and narrow in the end.

For the meck-via-Elixir, that would have worked if this test were running in the same VM but those tests are run separately against an instance of `dev/run`.

nickva commented on src/couch_views/src/couch_views_batch.erl:
```diff
@@ -0,0 +1,68 @@
+            ModAtom = list_to_existing_atom(ModStr),
+            {ModAtom, undefined}
+    end,
+    {NewState, BatchSize} = Mod:start(State),
```
Wonder if it makes sense to pass in dbname and/or ddoc id to the `Mod:start/1` so the batch maximizer could apply different rules to different databases or design docs perhaps. For example, like we pass in the DocsRead below even though we don't actually use it in the basic implementation.

davisp replied:
Good idea! I'll pass in the Mrst record to each of the three functions since its a public record that contains the database name and design document id.

nickva replied:
Good idea passing in the mrst record, even better than just the dbname and docid

nickva commented on src/couch_views/src/couch_views_indexer.erl:
```diff
@@ -78,10 +86,10 @@ init() ->
-
     State = #{
         tx_db => undefined,
+        tx_size => 0,
```
Minor nit: would these make sense emitted as metrics. It could a separate PR though

davisp replied:
I was noodling over a way to dynamically enable the tracing logs I've been using while developing this. As metrics I think the cardinality would either be high enough it'd crush most metric collectors or coarse enough that it wouldn't be super helpful.

nickva replied:
Yeah, makes sense. I was thinking of low cardinality ones just see a range of transaction sizes and estimated limits, but you're right it would bounce around too much and not be very useful.

iilyak commented:
> For the meck-via-Elixir, that would have worked if this test were running in the same VM but those tests are run separately against an instance of dev/run.

We have two kinds of Elixir tests
- integration tests which live in test/elixir/test
- unit tests which live in `src/<app>/test/exunit/`

Unit tests are run on the same VM

davisp commented:
@nickva I've addressed your three main suggestions. For the metrics I don't have a good answer other than maybe expose the timing logs I've currently somewhat hacked in the views branch. Let me know if you think some sort of combined metric there would be useful and I can add that.

nickva commented:
@davisp Looks, let's skip the metrics for now, could be a future enhancement

I am getting a consistent timeout in the `couch_views_indexer_test:39: indexer_test_ (indexed_empty_db)...` test while it passes on prototype/fdb-layer. Looking a bit into that

davisp commented:
@nickva I sometimes have one but it seems to be only occasionally for the first run after compiling and then doesn't happen again. Let me know if you find anything though. Flaky tests are never good.

nickva commented:
@davisp Found the issue, couch_views builds are were fast enough to trigger a race condition in couch_jobs monitoring code. The job was completing before the type monitor for couch_views jobs even started.

Made a PR to fix it:
  https://github.com/apache/couchdb/pull/3135

nickva commented on src/couch_views/src/couch_views_batch_impl.erl:
```diff
@@ -0,0 +1,147 @@
+        sense_incr = get_config("batch_sense_increment", "100"),
+        max_tx_size = get_config("batch_max_tx_size", "9000000"),
+        max_tx_time_msec = get_config("batch_max_tx_time_msec", "4500"),
+        threshold_penalty = get_config("batch_threshold_penalty", "0.2")
```
Minor nit: maybe set `state = search` explicitly here (as well or just here), just so all the state initialization is in one place and it's obvious that it always starts in `search`, stays there until it hits the threshold, then remains in `sense` after that.

davisp replied:
Done.

nickva commented on src/couch_views/src/couch_views_batch_impl.erl:
```diff
@@ -0,0 +1,147 @@
+        threshold_penalty = Penalty
+    } = St,
+    St#batch_st{
+        size = non_neg_integer(Size, batch_initial_size),
+        search_incr = non_neg_integer(SearchIncr, batch_search_increment),
+        sense_incr = non_neg_integer(SenseIncr, batch_sense_increment),
+        max_tx_size = non_neg_integer(MaxTxSize, batch_max_tx_size),
+        max_tx_time_msec = non_neg_integer(MaxTxTime, batch_max_tx_time_msec),
+        threshold_penalty = float_0_to_1(Penalty, batch_threshold_penalty)
```
Would this be shorter to combine with `get_config/2`. Maybe something like:

```
St = #batch_st{
    size = get_config(batch_initial_size, "100", fun non_neg_integer/2),
    search_incr = get_config(batch_search_increment, "500", fun non_neg_integer/2),
    ...
}
```

Then have 

```
get_config(Key, Default, ValFun) when is_function(ValFun, 2) ->
    Str = config:get("couch_views", atom_to_list(Key), Default),
    ValFun(Str, Key).
```

Up to you if you feel like it would eliminate the extra unpacking from validation...

davisp replied:
Done.

nickva commented on src/couch_views/test/couch_views_batch_test.erl:
```diff
@@ -0,0 +1,83 @@
+    couch_views_batch:start(#mrst{}),
+    couch_views_batch:success(#mrst{}, 0, 0),
+    ?assertEqual(150, couch_views_batch:start(#mrst{})).
+
```
Super tiny nit: extra newline

davisp replied:
Done.

nickva commented on src/couch_views/README.md:
```diff
@@ -7,42 +7,28 @@ Currently only map indexes are supported and it will always return the full inde
+; batch_search_increment = 500 ; Size change when searching for the threshold
+; batch_sense_increment = 100 ; Size change increment after hitting a threshold
+; batch_max_tx_size = 9000000 ; Maximum transaction size in bytes
+; batch_max_tx_time = 4500 ; Maximum transaction time in milliseconds
```
Update the default.ini config with the updated `batch_max_tx_time_msec` name?

davisp replied:
Done.

nickva reviewed (approved):
+1 Nice work!

And good comments from Ilya and Garren about re-structuring it as two modules.

Left a few style and param validation comments, up to you there if you feel they are too nitpicky

iilyak commented on src/couch_views/src/couch_views_batch.erl:
```diff
@@ -0,0 +1,80 @@
+-callback success(
+            Mrst::#mrst{},
+            TxSize::non_neg_integer(),
+            DocsRead::non_neg_integer(),
```
The `couch_rate:success/2` callback used `WrittenDocs` argument. Would it be possible to pass `DocsWrite` to the callback?

davisp replied:
If you look at the old implementation `WrittenDocs` is exactly the same value as the new `DocsRead`, except that `DocsRead` is taken directly from `length(DocAcc)` where as `WrittenDocs` is calculated while iterating over `DocAcc`.

iilyak replied:
My understanding is that map function can emit 0 or more documents, therefore `DocsRead /= DocsWritten`:

- DocsRead  - `length(DocAcc)`
- DocsWritten - `length(MappedDocs)`

In `couch_rate` the `WrittenDocs` was a result of iterating over `MappedDocs` https://github.com/apache/couchdb/blob/prototype/fdb-layer/src/couch_views/src/couch_views_indexer.erl#L205

The reason I calculated it during iteration is to avoid extra `O(n)` cost on `length(List)`.

davisp replied:
A map function can emit 0 or more rows for each of the defined views. For every doc we have a result set regardless of whether it has rows or not (because we have to remember we may be removing a document from a view). Regardless of whether a doc emits rows or not, it still has an entry in `MappedDocs` and `length(DocAcc) == length(MappedDocs)`.

I could see counting up the number of rows emitted for each document as maybe being an interesting value to have access to.

davisp replied:
I've tweaked the API to include the total number of emitted KVs and switched to using a map with a defined type exported from `couch_views_batch`. Let me know if you can think of any other useful values or if those three are good enough for now.

davisp commented:
@nickva I've addressed all of your suggestions. Let me know if you have anything else.

@iilyak @garrensmith I'm pretty sure I've addressed all of your concerns, let me know if there's anything left to handle.

nickva commented:
@davisp lgtm +1 to merge from me, also good job adding extra tests!

garrensmith commented on src/couch_views/README.md:
```diff
@@ -7,42 +7,28 @@ Currently only map indexes are supported and it will always return the full inde
 
 * `couch_views` - Main entry point to query a view
-* `couch_views_reader` - Reads from the index for queries
+* `couch_views_batch` - Dynamically determine optimal batch sizes for view indexers.
```
Can you add in `couch_views_batch_impl` here.

davisp replied:
Will do, good catch.

davisp replied:
Done.

garrensmith reviewed (changes requested):
One small change from me and then it looks good.

iilyak commented on rel/overlay/etc/default.ini:
```diff
@@ -312,6 +312,14 @@ iterations = 10 ; iterations for password hashing
+; batch_initial_size = 100 ; Initial batch size in number of documents
+; batch_search_increment = 500 ; Size change when searching for the threshold
+; batch_sense_increment = 100 ; Size change increment after hitting a threshold
+; batch_max_tx_size = 9000000 ; Maximum transaction size in bytes
```
Should we name it `batch_max_tx_size_bytes` to make it clear that units are different compared to `batch_initial_size`?

iilyak reviewed (approved):
+1. after `batch_max_tx_size -> batch_max_tx_size_bytes` rename.

garrensmith reviewed (approved):
+1

davisp commented:
Updated the max size parameter to include bytes. Will merge when CI comes back green.
